File: frontend/lib/logger.ts
/**
 * Structured JSON logger for production-ready logging.
 * Replaces console.log statements with structured, contextual logging.
 */

interface LogContext {
  [key: string]: string | number | boolean | null | undefined | object;
}

interface LogEntry {
  timestamp: string;
  level: string;
  logger: string;
  message: string;
  context?: LogContext;
}

type LogLevel = "DEBUG" | "INFO" | "WARN" | "ERROR";

class StructuredLogger {
  private name: string;
  private globalContext: LogContext = {};

  constructor(name: string) {
    this.name = name;
  }

  /**
   * Set global context that will be included in all log entries
   */
  setGlobalContext(context: LogContext): void {
    this.globalContext = { ...this.globalContext, ...context };
  }

  /**
   * Clear global context
   */
  clearGlobalContext(): void {
    this.globalContext = {};
  }

  /**
   * Internal logging method
   */
  private log(level: LogLevel, message: string, context?: LogContext): void {
    const logEntry: LogEntry = {
      timestamp: new Date().toISOString(),
      level,
      logger: this.name,
      message,
    };

    // Merge global context with local context
    const mergedContext = { ...this.globalContext, ...context };
    if (Object.keys(mergedContext).length > 0) {
      logEntry.context = mergedContext;
    }

    // In production, send to logging service
    if (process.env.NODE_ENV === "production") {
      this.sendToLoggingService(logEntry);
    } else {
      // Development: pretty print to console
      this.logToConsole(level, logEntry);
    }
  }

  /**
   * Log to console in development mode
   */
  private logToConsole(level: LogLevel, logEntry: LogEntry): void {
    const formattedLog = JSON.stringify(logEntry, null, 2);

    switch (level) {
      case "DEBUG":
        console.debug(formattedLog);
        break;
      case "INFO":
        console.info(formattedLog);
        break;
      case "WARN":
        console.warn(formattedLog);
        break;
      case "ERROR":
        console.error(formattedLog);
        break;
    }
  }

  /**
   * Send logs to external logging service in production
   * This can be integrated with services like Datadog, LogRocket, Sentry, etc.
   */
  private sendToLoggingService(logEntry: LogEntry): void {
    // TODO: Integrate with logging service
    // Example integrations:
    // - Datadog: datadogLogs.logger.log(logEntry.message, logEntry)
    // - LogRocket: LogRocket.log(logEntry)
    // - Sentry: Sentry.captureMessage(logEntry.message, { extra: logEntry.context })

    // For now, still log to console in production
    console.log(JSON.stringify(logEntry));
  }

  /**
   * Log debug message with optional context
   *
   * @param message - Debug message
   * @param context - Additional context (e.g., { userId: "123", page: "quiz" })
   */
  debug(message: string, context?: LogContext): void {
    this.log("DEBUG", message, context);
  }

  /**
   * Log info message with optional context
   *
   * @param message - Info message
   * @param context - Additional context (e.g., { userId: "123", action: "login" })
   */
  info(message: string, context?: LogContext): void {
    this.log("INFO", message, context);
  }

  /**
   * Log warning message with optional context
   *
   * @param message - Warning message
   * @param context - Additional context (e.g., { userId: "123", issue: "slow_response" })
   */
  warn(message: string, context?: LogContext): void {
    this.log("WARN", message, context);
  }

  /**
   * Log error message with optional context
   *
   * @param message - Error message
   * @param context - Additional context (e.g., { userId: "123", error: "api_timeout" })
   */
  error(message: string, context?: LogContext): void {
    this.log("ERROR", message, context);
  }
}

/**
 * Create a structured logger instance
 *
 * @param name - Logger name (typically the component or module name)
 * @returns StructuredLogger instance
 *
 * @example
 * const logger = createLogger('QuizPage');
 * logger.info('Quiz started', { userId: '123', topic: 'Python' });
 */
export function createLogger(name: string): StructuredLogger {
  return new StructuredLogger(name);
}

/**
 * Default logger instance for general use
 */
export const logger = createLogger("App");

File: backend/utils/logger.py
"""
Structured JSON logger for production-ready logging.
Replaces print() statements with structured, contextual logging.
"""

import logging
import json
import os
from datetime import datetime, timezone
from typing import Any, Dict, Optional


class JSONFormatter(logging.Formatter):
    """Custom JSON formatter for structured logging"""

    def format(self, record: logging.LogRecord) -> str:
        """Format log record as JSON"""
        log_data = {
            'timestamp': datetime.now(timezone.utc).isoformat(),
            'level': record.levelname,
            'logger': record.name,
            'message': record.getMessage(),
            'module': record.module,
            'function': record.funcName,
            'line': record.lineno
        }
        
        # Add extra context if provided
        if hasattr(record, 'context') and record.context:
            log_data['context'] = record.context
        
        # Add exception info if present
        if record.exc_info:
            log_data['exception'] = self.formatException(record.exc_info)
        
        return json.dumps(log_data)


class StructuredLogger:
    """
    Structured logger that outputs JSON-formatted logs.
    
    Usage:
        logger = StructuredLogger(__name__)
        logger.info("User logged in", user_id="123", action="login")
        logger.error("Failed to generate quiz", topic="Python", error="API timeout")
    """
    
    def __init__(self, name: str):
        """
        Initialize structured logger.
        
        Args:
            name: Logger name (typically __name__ of the module)
        """
        self.logger = logging.getLogger(name)
        
        # Configure log level from environment variable
        log_level = os.getenv('LOG_LEVEL', 'INFO').upper()
        self.logger.setLevel(getattr(logging, log_level, logging.INFO))
        
        # Avoid duplicate handlers
        if not self.logger.handlers:
            handler = logging.StreamHandler()
            handler.setFormatter(JSONFormatter())
            self.logger.addHandler(handler)
        
        # Prevent propagation to root logger
        self.logger.propagate = False
    
    def _log(self, level: int, message: str, **context: Any) -> None:
        """
        Internal logging method that adds context.
        
        Args:
            level: Logging level
            message: Log message
            **context: Additional context as keyword arguments
        """
        extra = {'context': context} if context else {}
        self.logger.log(level, message, extra=extra)
    
    def debug(self, message: str, **context: Any) -> None:
        """
        Log debug message with optional context.
        
        Args:
            message: Debug message
            **context: Additional context (e.g., user_id="123", topic="Python")
        """
        self._log(logging.DEBUG, message, **context)
    
    def info(self, message: str, **context: Any) -> None:
        """
        Log info message with optional context.
        
        Args:
            message: Info message
            **context: Additional context (e.g., user_id="123", action="login")
        """
        self._log(logging.INFO, message, **context)
    
    def warning(self, message: str, **context: Any) -> None:
        """
        Log warning message with optional context.
        
        Args:
            message: Warning message
            **context: Additional context (e.g., user_id="123", issue="rate_limit")
        """
        self._log(logging.WARNING, message, **context)
    
    def error(self, message: str, **context: Any) -> None:
        """
        Log error message with optional context.
        
        Args:
            message: Error message
            **context: Additional context (e.g., user_id="123", error="timeout")
        """
        self._log(logging.ERROR, message, **context)
    
    def exception(self, message: str, **context: Any) -> None:
        """
        Log exception with traceback and optional context.
        
        Args:
            message: Exception message
            **context: Additional context
        """
        extra = {'context': context} if context else {}
        self.logger.exception(message, extra=extra)


# Convenience function to create logger instances
def get_logger(name: str) -> StructuredLogger:
    """
    Get a structured logger instance.
    
    Args:
        name: Logger name (typically __name__)
    
    Returns:
        StructuredLogger instance
    """
    return StructuredLogger(name)
